index.js

```diff
@@ -3,55 +3,82 @@ const getIpfs = require('window.ipfs-fallback')
-    }
+  return {
+    name: 'ipfs',
 
-    if (type === 'IPFS_INIT_FAILED') {
-      return Object.assign({}, state, { ipfsReady: false, error: error })
-    }
+    persistActions: [
+      'IPFS_INIT_FINISHED',
+      'IPFS_API_UPDATED'
+    ],
 
-    if (type === 'IPFS_API_UPDATED') {
-      return Object.assign({}, state, { apiAddress: payload })
-    }
+    reducer (state, {type, payload, error}) {
+      state = state || defaultState
 
-    return state
-  },
+      if (type === 'IPFS_INIT_STARTED') {
+        return Object.assign({}, state, { error: null })
+      }
 
-  getExtraArgs () {
-    return { getIpfs: () => root.ipfs }
-  },
+      if (type === 'IPFS_INIT_FINISHED') {
+        ready = true
+        return Object.assign({}, state, { identity: payload })
+      }
 
-  selectIpfsReady: state => state.ipfs.ipfsReady,
+      if (type === 'IPFS_INIT_FAILED') {
+        ready = false
+        return Object.assign({}, state, { error: error })
+      }
 
-  selectIpfsIdentity: state => state.ipfs.identity,
+      if (type === 'IPFS_API_UPDATED') {
+        return Object.assign({}, state, { apiAddress: payload, error: null })
+      }
 
-  doInitIpfs: () => async ({ dispatch, getState }) => {
-    dispatch({ type: 'IPFS_INIT_STARTED' })
-    console.log('IPFS_INIT_STARTED')
-    const {apiAddress} = getState().ipfs
-    let identity = null
-    try {
-      root.ipfs = await getIpfs({ api: true, ipfs: apiAddress })
-      // will fail if remote api is not available on default port
-      identity = await root.ipfs.id()
-    } catch (error) {
-      return dispatch({ type: 'IPFS_INIT_FAILED', error })
-    }
-    dispatch({ type: 'IPFS_INIT_FINISHED', payload: identity })
-    console.log('IPFS_INIT_FINISHED')
-  },
+      return state
+    },
+
+    getExtraArgs () {
+      return { getIpfs: () => root.ipfs }
+    },
+
+    selectIpfsReady: () => ready,
+
+    selectIpfsApiAddress: state => state.ipfs.apiAddress,
+
+    selectIpfsInitFailed: state => !!state.ipfs.error,
 
-  doUpdateIpfsAPIAddress: (apiAddress) => ({dispatch, store}) => {
-    dispatch({type: 'IPFS_API_UPDATED', payload: apiAddress})
-    store.doInitIpfs()
+    selectIpfsIdentity: state => state.ipfs.identity,
+
+    doInitIpfs: () => async ({ dispatch, getState }) => {
+      dispatch({ type: 'IPFS_INIT_STARTED' })
+
+      const { apiAddress } = getState().ipfs
+      let identity = null
+
+      try {
+        root.ipfs = await getIpfs({ api: true, ipfs: apiAddress })
+        // will fail if remote api is not available on default port
+        identity = await root.ipfs.id()
+      } catch (error) {
+        return dispatch({ type: 'IPFS_INIT_FAILED', error })
+      }
+
+      dispatch({ type: 'IPFS_INIT_FINISHED', payload: identity })
+    },
+
+    doUpdateIpfsAPIAddress: (apiAddress) => ({dispatch, store}) => {
+      // root.ipfs needs to be set to null so getIpfs doesn't
+      // use the previous connection.
+      root.ipfs = null
```

Comment:
`root.ipfs` could be the `window.ipfs`  instance from companion. Now this bundle exports a function, we should store the current ipfs instance as a variable in the function scope, and not mess with the `root.ipfs` instance.